### DISCUSSION TITLE
[Generic] Inference decomposition

### DISCUSSION THREAD

**Discussion by notour** [dotnet/csharplang#7151] [Language Ideas] [answered]
Take the following example :

        public interface IStage<TInput, TOuput>
        {
        }

        public class SampleStage : IStage<string, Guid>
        {

        }

        public class Builder
        {
            public Builder AddStage<TInput, TOutput>(IStage<TInput, TOutput>)
            {
                throw new NotImplementedException();
            }

            public Builder AddStage<TStage, TInput, TOutput>()
                where TStage : IStage<TInput, TOutput>
            {
                throw new NotImplementedException();
            }
        }

        public class IT
        {
            public static T Get<T>()
            {
                return default;
            }
        }

        public void Test()
        {
            var builder = new Builder();

            // Work but require to provide an instance
            builder.AddStage(new SampleStage());

            // Work around but kind of heavy to right
            builder.AddStage(IT.Get<SampleStage>();

            // Work but had to specify all parameter
            builder.AddStage<SampleStage, string, Guid>();

            // WOULD LIKE
            builder.AddStage<SampleStage>();

            // OR WOULD LIKE
            builder.AddStage<SampleStage, ...>();
        }

C# could solve through inference of type if you provide the final type in argument.

It would be wonderfull to resolve partally the generic parameter and based on "where" condition detect automatically the others.
This will simply fluent syntax and compact some generic usage.

Syntax proposed : 

            // WOULD LIKE
            builder.AddStage<SampleStage>();

            // OR WOULD LIKE
            builder.AddStage<SampleStage, ...>();

            // WOULD LIKE
            builder.AddStage<<SampleStage>>();

**Comment by 333fred** [comment, marked as answer]
This is really a combination of https://github.com/dotnet/csharplang/issues/1349 and https://github.com/dotnet/csharplang/discussions/741.
